easybuild/easyblocks/n/nvhpc.py

```diff
@@ -0,0 +1,201 @@
+#
+# This file is triple-licensed under GPLv2 (see below), MIT, and
+# BSD three-clause licenses.
+#
+# This file is part of EasyBuild,
+# originally created by the HPC team of Ghent University (http://ugent.be/hpc/en),
+# with support of Ghent University (http://ugent.be/hpc),
+# the Flemish Supercomputer Centre (VSC) (https://www.vscentrum.be),
+# Flemish Research Foundation (FWO) (http://www.fwo.be/en)
+# and the Department of Economy, Science and Innovation (EWI) (http://www.ewi-vlaanderen.be/en).
+#
+# https://github.com/easybuilders/easybuild
+#
+# EasyBuild is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation v2.
+#
+# EasyBuild is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with EasyBuild.  If not, see <http://www.gnu.org/licenses/>.
+##
+"""
+EasyBuild support for installing NVIDIA HPC SDK compilers, based on the easyblock for PGI compilers
+
+@author: Bart Oldeman (McGill University, Calcul Quebec, Compute Canada)
+@author: Damian Alvarez (Forschungszentrum Juelich)
+@author: Andreas Herten (Forschungszentrum Juelich)
+"""
+import os
+import fileinput
+import re
+import stat
+import sys
+
+import easybuild.tools.environment as env
+from distutils.version import LooseVersion
+from easybuild.easyblocks.generic.packedbinary import PackedBinary
+from easybuild.framework.easyconfig import CUSTOM
+from easybuild.tools.filetools import adjust_permissions, write_file
+from easybuild.tools.run import run_cmd
+from easybuild.tools.modules import get_software_root
+
+
+# contents for siterc file to make PGI/NVHPC pick up $LIBRARY_PATH
+# cfr. https://www.pgroup.com/support/link.htm#lib_path_ldflags
+SITERC_LIBRARY_PATH = """
+# get the value of the environment variable LIBRARY_PATH
+variable LIBRARY_PATH is environment(LIBRARY_PATH);
+
+# split this value at colons, separate by -L, prepend 1st one by -L
+variable library_path is
+default($if($LIBRARY_PATH,-L$replace($LIBRARY_PATH,":", -L)));
+
+# add the -L arguments to the link line
+append LDLIBARGS=$library_path;
+
+# also include the location where libm & co live on Debian-based systems
+# cfr. https://github.com/easybuilders/easybuild-easyblocks/pull/919
+append LDLIBARGS=-L/usr/lib/x86_64-linux-gnu;
+"""
+
+
+class EB_NVHPC(PackedBinary):
+    """
+    Support for installing the NVIDIA HPC SDK (NVHPC) compilers
+    """
+
+    @staticmethod
+    def extra_options():
+        extra_vars = {
+            'default_cuda_version': ["10.1", "CUDA Version to be used as default (10.2, 11.0, ...)", CUSTOM],
+            'compute_capability':   ["70", "Compute Capability (70, 80, ...)", CUSTOM],
```

Comment:
Does NVHPC support installing for a set compute capabilities too?

If so, we should leverage the top-level `cuda_compute_capabilities` easyconfig parameter instead (introduced in #3382)

Reply:
The options are given further to the NVHPC `install` script which, in turn, writes it to a `cudarc` file, similarly to the old PGI RC files. The contents after calling the EB block with 11.0 and 70 are
```
set DEFCUDAVERSION=11.0;
set DEFSTDPARCOMPUTECAP=70;
```

I don't know if setting a set of compute capabilities is supported, but my guess would be: yes. 

If the overriding `$NVHPC_STDPAR_CUDACC` (which we use) is not set, `install` tries to determine the compute capability itself by calling `nvaccelinfo -dev 0 | grep -i "default target" | cut -d":" -f2 | tr -d 'c\040\015\012\011'`; resulting in `70` on a V100 system.

Finally, to your question: I guess we could use `cuda_compute_capabilities`? How would that be specified in the easyconfig?

Reply:
In the easyconfigs it works like this: `cuda_compute_capabilities = ['3.0', '3.2', '3.5', '3.7', '5.0', '5.2', '5.3', '6.0', '6.1', '7.0']`

I guess one could get rid of the `compute_capability` parameter here, and simply use `cuda_compute_capabilities` in the easyblock, converted as `','.join(''.join(x.split('.')) for x in cuda_compute_capabilities)` or something similar? I think keeping it consistent with other packages makes sense.

Could we give it a quick try @AndiH ?

Reply:
Indeed, it would be confusing to have both `cuda_compute_capabilities` (which is a general easyconfig parameter since https://github.com/easybuilders/easybuild-framework/pull/3382) and `compute_capability`, especially if the value set for `cuda_compute_capabilities` is blatantly ignored.

This easyblock should also take into account the value of the `--cuda-compute-capabilities` EasyBuild configuration option, like we do elsewhere (see https://github.com/easybuilders/easybuild-easyblocks/blob/5c8c30c5d9f86cf4fa82b7d5befd7af90aff74d9/easybuild/easyblocks/t/tensorflow.py#L518-L523 for example)

Reply:
Alright. I'm not super convinced (because using `cuda_compute_capabilities` suggests that something is compiled for this Compute Capability, which it is not – it is merely written as the default for any NVHPC-based compilations for the future), but I'll implement.

Reply:
I poked (well `rg`-grepped around) and found this note:
```
# Currently, stdpar is limited to a single compute capability, so it needs its
# own special case for detection and default. Once we can support multiple
# compute capabilites we can remove all of the *STDPARCOMPUTECAP* variables and
# only use COMPUTECAP like we do for all other GPU codes.
```

I still would implement the block such that it is using `cuda_compute_capabilities` but would make a note (where?) that only single values are supported.

Reply:
(I've also learned something new. `DEFSTDPARCOMPUTECAP` is indeed only used for the pSTL (as the name suggests) and if one were to set a general default for the compute capability, one should rather choose `COMPUTECAP` (which, probably, also goes to  `cudarc`).)

Reply:
`--cuda-compute-capabilities` accepts only values with a dot, for example `7.0`. But that is not something understood by NVHPC and also not how you would specify it for any of the compilers in the CLI, for example `nvc -ta=tesla:cc70`.
I'll add the dot-removal now…
(But I'm getting less and less convinced that this is the right thing to do.)

Reply:
Implemented in 2356db58